apps/minifront/src/components/shared/asset-selector.tsx

```diff
@@ -0,0 +1,126 @@
+import {
+  Dialog,
+  DialogClose,
+  DialogContent,
+  DialogHeader,
+  DialogTrigger,
+} from '@penumbra-zone/ui/components/ui/dialog';
+import { AssetIcon } from '@penumbra-zone/ui/components/ui/tx/view/asset-icon';
+import {
+  Metadata,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import { localAssets } from '@penumbra-zone/constants/src/assets';
+import { ValueViewComponent } from '@penumbra-zone/ui/components/ui/tx/view/value';
+import { useEffect, useMemo, useState } from 'react';
+import { IconInput } from '@penumbra-zone/ui/components/ui/icon-input';
+import { MagnifyingGlassIcon } from '@radix-ui/react-icons';
+
+interface AssetSelectorProps {
+  value?: Metadata;
+  onChange: (metadata: Metadata) => void;
+  /**
+   * If passed, this function will be called for every asset that
+   * `AssetSelector` plans to display. It should return `true` or `false`
+   * depending on whether that asset should be displayed.
+   */
+  filter?: (metadata: Metadata) => boolean;
+}
+
+const sortedAssets = [...localAssets].sort((a, b) =>
+  a.symbol.toLocaleLowerCase() < b.symbol.toLocaleLowerCase() ? -1 : 1,
+);
+
+/**
+ * If the `filter` rejects the currently selected `asset`, switch to a different
+ * `asset`.
+ */
+const switchAssetIfNecessary = ({
+  value,
+  onChange,
+  filter,
+  assets,
+}: AssetSelectorProps & { assets: Metadata[] }) => {
+  if (!filter || !value) return;
+
+  if (!filter(value)) {
+    const firstAssetThatPassesTheFilter = assets.find(filter);
+    if (firstAssetThatPassesTheFilter) onChange(firstAssetThatPassesTheFilter);
+  }
+};
+
+const useFilteredAssets = ({ value, onChange, filter }: AssetSelectorProps) => {
+  const [search, setSearch] = useState('');
```

Comment:
This search functionality is new — it used to only be in `<SelectTokenModal />` (which is now renamed to `<BalanceSelector />`).